jax/_src/lax/slicing.py

```diff
@@ -222,7 +220,7 @@ def from_any(s: Optional[Union[str, 'GatherScatterMode']]):
       raise ValueError(f'Unknown gather mode "{s}"')
 
 
-def gather(operand: Array, start_indices: Array,
+def gather(operand: ArrayLike, start_indices: ArrayLike,
```

Comment:
Aren't `operand` and `start_indices` always of type `Array`? Same question for `scatter` etc. IMO, it would be quite surprising otherwise

Reply:
`operand` here is passed to  `_dtype` and to `Primitive.bind`, neither of which require inputs of type `Array`, but do accept the union of types that we call `ArrayLike`.

Reply:
I mean the doc string does mention `operand` being an array. And I would argue that the semantic of `gather` etc. seem to imply the first two arguments to be arrays (not scalars, as `jac.ArrayLike` already ruled out types where canonical projections into `ndarray` are defined). For instance, if I can make a call `gather(m, n, ...)` where `m` and `n` are integers, I'd say it's the interface being too forgiving.

Anyhow, thanks for explaining.

Reply:
Well, `Array` would also rule-out `np.ndarray`, a valid input to this function, while `ArrayLike` does not.

It would be impossible in Python's current implementation to make type annotations exactly match the set of valid inputs; for example `jnp.array(0)` is of type `Array` but has the wrong shape. Similarly, the scalar value `np.int32(0)` is able to be converted implicitly to `Array` (which is why it is `ArrayLike`) but has the wrong shape.

There's no perfect type annotation, but all told I think `ArrayLike` is the correct annotation here.

Reply:
By the way, the principle I'm following here is [inputs should be permissively typed](https://jax.readthedocs.io/en/latest/jep/12049-type-annotations.html#inputs-should-be-permissively-typed), but I agree there's wiggle room and it's always a judgment call.